kernel/mempool.c

```diff
@@ -40,6 +40,37 @@ static void *z_heap_aligned_alloc(struct k_heap *heap, size_t align, size_t size
 	return mem;
 }
 
+static void *z_heap_aligned_realloc(struct k_heap *heap, void *ptr, size_t align, size_t size)
+{
+	void *mem;
+	struct k_heap **heap_ref;
+	size_t __align;
+
+	/*
+	 * Adjust the size to make room for our heap reference.
+	 * Merge a rewind bit with align value (see sys_heap_aligned_realloc()).
+	 * This allows for storing the heap pointer right below the aligned
+	 * boundary without wasting any memory.
+	 */
+	if (size_add_overflow(size, sizeof(heap_ref), &size)) {
+		return NULL;
+	}
+	__align = align | sizeof(heap_ref);
```

Comment:
Similar comment: two boilerplate functions would be nice to unify.  But more important here because this is actually slightly non-trivial logic to muck with the allocation layout.  Maybe factor out a "adjust_heap_size()" routine before the allocation and a "add_heap_ref()" routine after to avoid the cut/paste.